### PR TITLE
Broadcasting to multiple rooms causes to receive a message more than once

### PR DESCRIPTION
**Pull request by nkzawa**
(This PR doesn't solve the problem yet)

The problem happens when a socket joins to multiple rooms then a message is sent to these rooms at a time.
Please see the test case for how to reproduce.

This should be a bug since it looks socket.io-adapter doesn't work like this.
https://github.com/socketio/socket.io-adapter/blob/master/index.js#L132 

The root cause is that we have to publish a message to each room in loop, so that a node which received the message can't figure out whether it already accepted or not. 
https://github.com/socketio/socket.io-redis/blob/master/index.js#L130

### CODE REVIEW

**Comment by nkzawa** [comment]
I think it's not ideal, but added fix for the issue anyway.

File: index.js [hunk 1]
```diff
@@ -238,6 +261,14 @@ function adapter(uri, opts){
     });
   };
 
+  /**
+   * Closes and clean up the adapter.
+   */
+
+  Redis.prototype.close = function(){
+    clearInterval(this.cleanupIntervalId);
```
Comment (nkzawa) [inline index.js]: We'd have to call this in somewhere appropriate.

**Comment by rauchg** [comment]
We should discuss this more.

**Comment by nkzawa** [comment]
Definitely. I'm still suspicious about subscribing each room.

**Comment by nkzawa** [comment]
As another solution, we can just publish a message to `#nsp#` channel rather than `#nsp#room#` if the message has multiple rooms as target, like the following.

```js
  var chn = prefix + '#' + packet.nsp + '#';
  var msg = msgpack.encode([uid, packet, opts]);
  if (opts.rooms && opts.rooms.length === 1) {
    pub.publish(chn + opts.rooms[0] + '#';, msg);
  } else {
    pub.publish(chn, msg);
  }
```

**Comment by darrachequesne** [comment]
@nkzawa I think that last solution should work well! The multiplicity of rooms is handled in the method `Adapter.prototype.broadcast`, right?

The other one involving periodic cleanups seems a bit too complicated to me.

**Comment by nkzawa** [comment]
@darrachequesne agree, but this PR would be faster in some circumstance, like there are a lot of nodes which have really sparse rooms. Anyway I also prefer the last solution :)

**Pull request closed without merging**